File: backend/throughput_reliability.py
from __future__ import annotations

import math
from collections.abc import Iterable, Sequence

from .simulation_value_objects import StatisticalValueError, ThroughputReliability


def _linear_quantile(sorted_values: Sequence[float], level: float) -> float:
    position = (len(sorted_values) - 1) * level
    lower_index = math.floor(position)
    upper_index = min(lower_index + 1, len(sorted_values) - 1)
    weight = position - lower_index
    lower_value = sorted_values[lower_index]
    return lower_value + weight * (sorted_values[upper_index] - lower_value)


def _strict_samples(samples: Iterable[object]) -> tuple[int, ...]:
    try:
        raw_values = tuple(samples)
    except TypeError as exc:
        raise StatisticalValueError("throughput_samples doit etre une collection.") from exc
    if not raw_values:
        raise StatisticalValueError("throughput_samples est vide.")
    if any(type(value) is not int or value < 0 for value in raw_values):
        raise StatisticalValueError(
            "throughput_samples doit contenir uniquement des entiers finis >= 0."
        )
    return raw_values


def calculate_throughput_reliability(
    samples: Iterable[object],
) -> ThroughputReliability:
    values = _strict_samples(samples)
    samples_count = len(values)
    mean = sum(values) / samples_count
    variance = sum((value - mean) ** 2 for value in values) / samples_count
    sorted_values = sorted(values)
    q25 = _linear_quantile(sorted_values, 0.25)
    median = _linear_quantile(sorted_values, 0.5)
    q75 = _linear_quantile(sorted_values, 0.75)

    mean_x = (samples_count - 1) / 2
    slope_denominator = sum((index - mean_x) ** 2 for index in range(samples_count))
    slope_numerator = sum((index - mean_x) * (value - mean) for index, value in enumerate(values))
    slope = slope_numerator / slope_denominator if slope_denominator > 0 else 0.0
    return ThroughputReliability.create(
        cv=math.sqrt(variance) / mean if mean > 0 else 0.0,
        iqr_ratio=(q75 - q25) / median if median > 0 else 0.0,
        slope_norm=slope / mean if mean > 0 else 0.0,
        samples_count=samples_count,
        mean=mean,
    )

File: frontend/src/domain/throughputReliability.ts
import type { ThroughputReliability } from "./simulationValueObjects";
import { createThroughputReliability } from "./simulationValueObjects";

function linearQuantile(sortedValues: readonly number[], level: number): number {
  const position = (sortedValues.length - 1) * level;
  const lowerIndex = Math.floor(position);
  const upperIndex = Math.min(lowerIndex + 1, sortedValues.length - 1);
  const weight = position - lowerIndex;
  const lowerValue = sortedValues[lowerIndex] ?? 0;
  const upperValue = sortedValues[upperIndex] ?? lowerValue;
  return lowerValue + weight * (upperValue - lowerValue);
}

export function computeThroughputReliability(
  samples: readonly number[],
): ThroughputReliability | null {
  if (samples.length === 0) return null;
  if (samples.some((value) => !Number.isSafeInteger(value) || value < 0)) {
    throw new Error("throughput_samples doit contenir uniquement des entiers finis >= 0.");
  }
  const values = [...samples];
  const samplesCount = values.length;
  const mean = values.reduce((sum, value) => sum + value, 0) / samplesCount;
  const variance = values.reduce(
    (sum, value) => sum + (value - mean) ** 2,
    0,
  ) / samplesCount;
  const sortedValues = [...values].sort((left, right) => left - right);
  const q25 = linearQuantile(sortedValues, 0.25);
  const median = linearQuantile(sortedValues, 0.5);
  const q75 = linearQuantile(sortedValues, 0.75);

  const meanX = (samplesCount - 1) / 2;
  const slopeNumerator = values.reduce(
    (sum, value, index) => sum + (index - meanX) * (value - mean),
    0,
  );
  const slopeDenominator = values.reduce(
    (sum, _value, index) => sum + (index - meanX) ** 2,
    0,
  );
  const slope = slopeDenominator > 0 ? slopeNumerator / slopeDenominator : 0;
  return createThroughputReliability({
    cv: mean > 0 ? Math.sqrt(variance) / mean : 0,
    iqrRatio: median > 0 ? (q75 - q25) / median : 0,
    slopeNorm: mean > 0 ? slope / mean : 0,
    samplesCount,
    mean,
  });
}
